### DISCUSSION TITLE
Should we add hooks?

### DISCUSSION THREAD

**Discussion by andreivladbrg** [sablier-labs/flow#33] [Ideas]
A useful feature for lockup streams is the hooks.

In the context of open-ended, I personally believe they wouldn't be that helpful since:
- the product is oriented towards the sender
- we don't have NFTs

I am opening this discussion to see if you have other opinions.

@sablier-labs/solidity

**Comment by PaulRBerg** [comment]
I disagree. I think that hooks would be helpful.

- A recipient may want to be notified when a sender tops up their negative balance stream
- A sender may want to know when a recipient claims their token, e.g. for market sentiment analysis purposes

P.S. why don't we have NFTs?

**Reply by razgraf** [reply]
I've opened https://github.com/sablier-labs/v2-open-ended/discussions/34 for the NFT question.

**Reply by PaulRBerg** [reply]
cheers

**Comment by andreivladbrg** [comment]
Hmm, let's brainstorm more about the possibilities before we make a final decision.

What are the functions you have in mind for adding the hooks, and for whom? (e.g. in the `deposit` function run only the recipient hook)

**Reply by PaulRBerg** [reply]
Could you take a stab at designing a spec yourself?

**Comment by smol-ninja** [comment]
> A recipient may want to be notified when a sender tops up their negative balance stream
A sender may want to know when a recipient claims their token, e.g. for market sentiment analysis purposes

Ain't these two "events" based notifications? I don't understand how hooks are exactly useful here.

However, I think hooks can have other use cases such as cross-contract actions. For examples,

- Some sort of reward calculation when a recipient withdraws from the stream.
- The sender implements a hook to top up the stream by checking the amount every time a recipient withdraws.

There could be many more that I can't anticipate right now but hooks are powerful features.

As for the design:
- `onStreamWithdrawn` on both Sender and Recipient
- `onStreamCancel` on Recipient
- `onStreamRefund` on Recipient
- `onAdjustRarePerSecond` on Recipient

Any thoughts on above hook endpoints? If we agree, I will create issues.

**Reply by PaulRBerg** [reply]
The point is to get notified on-chain, e.g., have a Safe multisig immediately respond to an action performed in a stream. I agree that the demand for this should be relatively little, though.

> other use cases such as cross-contract action
...
>  hooks are powerful features.

Indeed!

> Any thoughts on above hook endpoints?

I am not familiar with the API (yet), but the list sounds goods to me.

**Reply by smol-ninja** [reply]
Cool. I will create the issue for now. We can always change it later before the first release.

**Comment by smol-ninja** [comment]
Issue:
- https://github.com/sablier-labs/v2-open-ended/issues/45

**Comment by andreivladbrg** [comment]
I changed my mind, we should have hooks.

**Reply by PaulRBerg** [reply]
my favorite words

**Reply by andreivladbrg** [reply]
haha 😅
